astropy/io/ascii/tests/test_c_reader.py

```diff
@@ -888,3 +888,101 @@ def test_fast_tab_with_names(parallel, read_tab):
+                     names=('a', 'b', 'c', 'd'))
+    table = ascii.read(text, format='basic', guess=False,
+                       fast_reader={'parallel': parallel, 'exponent_style': 'D'})
+    assert_table_equal(table, expected)
+
+@pytest.mark.parametrize("parallel", [
+    pytest.mark.xfail(os.name == 'nt', reason=
+                      "Multiprocessing is currently unsupported on Windows")(True),
+    False])
+def test_fortran_reader(parallel):
+    """
+    Make sure that ascii.read() can read Fortran-style exponential notation
+    using the fast_reader.
+    """
+    text = 'A B C\n100.01D+99 2.0 3\n4.2d-1 5.0D-1 0.6D4'
+    expected = Table([[1.0001e101, 0.42], [2, 0.5], [3.0, 6000]],
+                     names=('A', 'B', 'C'))
+
+    # C strtod (not-fast converter) can't handle Fortran exp
+    with pytest.raises(FastOptionsError) as e:
+        ascii.read(text, format='basic', guess=False, fast_reader=
+                   {'parallel': parallel, 'use_fast_converter': False,
+                    'exponent_style': 'D'})
+    assert 'fast_reader: exponent_style requires use_fast_converter' in str(e)
+
+    # Enable multiprocessing and the fast converter
+    table = ascii.read(text, format='basic', guess=False,
+                       fast_reader={'parallel': parallel, 'exponent_style': 'D'})
+    assert_table_equal(table, expected)
+
+    text = 'A B C\n0.10001Q102 20.0Q-1 3\n.42q0 0.5 6.Q3'
+    table = ascii.read(text, format='basic', guess=False,
+                       fast_reader={'parallel': parallel, 'exponent_style': 'q'})
+    assert_table_equal(table, expected)
+
+    # mixes and triple-exponents without any character using autodetect option
+    text = 'A B C\n1.0001+101 2.0E0 3\n.42d0 0.5 0.42-123'
+    expected['C'][1] = 4.2e-124
+    table = ascii.read(text, format='basic', guess=False,
+                       fast_reader={'parallel': parallel, 'exponent_style': 'fortran'})
+    assert_table_equal(table, expected)
+
+    # additional corner-case checks
+    text = 'A B C\n1.0001+101 2.0+000 3\n0.42+000 0.5 .42-000'
+    expected['C'][1] = 4.2e-1
+    table = ascii.read(text, format='basic', guess=False,
+                       fast_reader={'parallel': parallel, 'exponent_style': 'fortran'})
+    assert_table_equal(table, expected)
+
+
+@pytest.mark.parametrize("parallel", [
+    pytest.mark.xfail(os.name == 'nt', reason=
+                      "Multiprocessing is currently unsupported on Windows")(True),
+    False])
+def test_fortran_invalid_exp(parallel):
+    """
+    Test Fortran-style exponential notation in the fast_reader with invalid
+    exponent-like patterns (no triple-digits) to make sure they are returned
+    as strings instead, as with the standard C parser.
+    """
+    if TRAVIS:
+        pytest.xfail("Large exponents can sometimes fail on Travis CI")
+
+    text = 'A B C D E\n1.0001+1 2.3+10 3+1001 2 8.e3\n.42d0 0.5 3 4.56-123.4 0.42-123'
```

Comment:
Testing with two rows is not so definitive because if either value parses as a string then the whole column is converted to string.  Better to write a clean, easy to read test like:
```
values = ['1.0001+1', '2.3+10', ..., '0.42-123']
expected = ['1.0001+1', '2.3+10', ..., 4.2e-122]
table = ascii.read(' '.join(values), format='no_header', ...)
out_values = [col[0] for col in table.itercols()]
assert out_values == expected
```